compiler/noirc_frontend/src/hir_def/types.rs

```diff
@@ -439,21 +439,42 @@ pub enum TypeVariableKind {
     /// Can bind to any type
     Normal,
 
+    /// A polymorphic type.
+    Polymorphic(PolymorphicKind),
+
+    /// A potentially constant array size. This will only bind to itself or
+    /// Type::Constant(n) with a matching size. This defaults to Type::Constant(n) if still unbound
+    /// during monomorphization.
+    Constant(u32),
```

Comment:
Unrelated to this PR: I think this variant is unused. I should make an issue for removing it.